ktor-server/ktor-server-plugins/ktor-server-call-logging/jvm/src/io/ktor/server/plugins/callloging/CallLogging.kt

```diff
@@ -5,265 +5,92 @@
-        override fun install(pipeline: Application, configure: Configuration.() -> Unit): CallLogging {
-            val loggingMonitoringPhase = PipelinePhase("LoggingMonitoringCall")
-            val loggingBeforeCallPhase = PipelinePhase("LoggingBeforeCall")
-            val loggingAfterCallPhase = PipelinePhase("LoggingAfterCall")
-            val configuration = Configuration().apply(configure)
-            val plugin = CallLogging(
-                configuration.logger ?: pipeline.log,
-                pipeline.environment.monitor,
-                configuration.level,
-                configuration.filters.toList(),
-                configuration.mdcEntries.toList(),
-                configuration.formatCall
-            )
-
-            pipeline.insertPhaseBefore(ApplicationCallPipeline.Monitoring, loggingMonitoringPhase)
-            pipeline.insertPhaseBefore(ApplicationCallPipeline.Call, loggingBeforeCallPhase)
-            pipeline.insertPhaseAfter(ApplicationCallPipeline.Fallback, loggingAfterCallPhase)
-
-            if (plugin.mdcEntries.isNotEmpty()) {
-                pipeline.intercept(loggingMonitoringPhase) {
-                    plugin.withMDC(call) {
-                        proceed()
-                    }
-                }
-                pipeline.intercept(loggingBeforeCallPhase) {
-                    plugin.withMDC(call) {
-                        proceed()
-                    }
-                }
-                pipeline.intercept(loggingAfterCallPhase) {
-                    plugin.withMDC(call) {
-                        proceed()
-                        plugin.logSuccess(call)
-                    }
-                }
-            } else {
-                pipeline.intercept(loggingMonitoringPhase) {
-                    proceed()
-                    plugin.logSuccess(call)
-                }
-            }
+    logCallsWithMDC(::logSuccess)
+}
 
-            return plugin
+private fun ApplicationPluginBuilder<CallLoggingConfig>.logCompletedCalls(logSuccess: (ApplicationCall) -> Unit) {
+    onCall { call ->
+        call.afterFinish {
+            if (it == null) return@afterFinish
+            logSuccess(call)
         }
     }
+}
 
-    public override suspend fun withMDCBlock(call: ApplicationCall, block: suspend () -> Unit) {
-        withMDC(call, block)
-    }
+private fun ApplicationPluginBuilder<CallLoggingConfig>.logCallsWithMDC(logSuccess: (ApplicationCall) -> Unit) {
+    val entries = pluginConfig.mdcEntries
 
-    /**
-     * Invoke suspend [block] with a context having MDC configured.
-     */
-    private suspend inline fun withMDC(call: ApplicationCall, crossinline block: suspend () -> Unit) {
-        withContext(MDCSurvivalElement(setupMdc(call))) {
-            try {
-                block()
-            } finally {
-                cleanupMdc()
-            }
-        }
+    on(MDCHook(ApplicationCallPipeline.Monitoring)) { call, block ->
```

Comment:
Nit: here and below --  this code will be executed each time one installs your plugin.
Hence, new objects will be created. Maybe creating static objects inside `MDCHook` will be just a bit more efficient :)